### DISCUSSION TITLE
Is there any way to add referrer header?

### DISCUSSION THREAD

**Discussion by Dave-Jalpan** [MicrosoftEdge/WebView2Feedback#3883] [Q&A]
I want to open a new tab instead of new window. But while doing that one of the application throws exception "Error: No referrer", as application is expecting referrer value in header. So how I can pass referrer in header from e.NewWindow while redirecting new window request to new tab? 

```
private void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
{
    baseTemplate.AddTab(e.Uri, true); /// Creating new tab instead opening new window. 
    e.Handled = true;
}
```

**Comment by Symbai** [comment]
Show us the code of `baseTemplate.AddTab(e.Uri, true);`

**Comment by Dave-Jalpan** [comment]
Hi @Symbai 

baseTemplate.AddTab is custom way to create tabs.

AddTab has following implementation 

```
CustomTabItem newTab = (CustomTabItem)GenerateNewItem(); /// Create tab item 
newTab.Content = newTab.GenerateWebBrowser(url);  // Create WebBrowser (WebView2) and attach to newly created tab item
currentTabControl.AddTab(newTab, isSelected);  // Attach new tab item to tab control. 
```

**Comment by Symbai** [comment]
I guess its because you create a new webview2 and visit the url (which then has no referrer). No tested by you can try using the `NewWindow` property on the `NewWindowRequested` event to obtain the webview2 and use this instead of creating a new one. Otherwise use the `GetDeferral` method, listen to  `WebResourceRequested` event, obtain the headers. Create a new tab with a new webview2 control, listen to its `WebResourceRequested` event and then set the headers to the previously obtained headers. I'm just not sure if `WebResourceRequested` would get raised before or after `NewWindowRequested` event raises.

Anyway to answer your original question, changing http header is done by `WebResourceRequested` event.

**Comment by Dave-Jalpan** [comment]
@Symbai 

Sorry, but is there any code sample which I can follow?
Which opens in new tab rather new window and maintain headers.

**Comment by ukandrewc** [comment]
@Symbai @Dave-Jalpan If you use `GetDeferral`, you won't need `WebResourceRequested` as there will be a referrer header.

@Dave-Jalpan You've already had a sample here: #873

**Comment by Dave-Jalpan** [comment]
@ukandrewc 
As mentioned when I converted your code I was not able to get newwindow deferrer 
So please correct me, and explain how to do it.

**Comment by ukandrewc** [comment]
@Dave-Jalpan Sorry, I can't help any more, I did answer that question, but I'm very rusty on C++, so can't help with that.

**Comment by Dave-Jalpan** [comment]
@ukandrewc 

Is this interpretation correct from  vb to c#

Friend Deferral As CoreWebView2Deferral
Friend EventArgs As CoreWebView2NewWindowRequestedEventArgs 
as

```
internal Microsoft.Web.WebView2.Core.CoreWebView2Deferral Deferral;
internal Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs EventArgs;

```

and Dim Browser As New MyWebView2
as

```
Microsoft.Web.WebView2.Wpf.WebView2 Browser = new Microsoft.Web.WebView2.Wpf.WebView2();
```

if so then i am not able to set deferral 

```
Browser.Deferral = = e.GetDeferral;
```

**Comment by ukandrewc** [comment]
@Dave-Jalpan Internals are correct but MyWebView2, needs to be your own inherited class

```
internal partial class MyWebView2 : WebView2
{
    internal Global.Microsoft.Web.WebView2.Core.CoreWebView2Deferral Deferral;
    internal Global.Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs EventArgs;
}
```

**Comment by Dave-Jalpan** [comment]
@ukandrewc 
How I can register close request event for newly open window from CoreWebView2_NewWindowRequested
I am having difficulty to handle newly open window.  

```
private void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
        {
  // e.NewWindow is null
            e.NewWindow.WindowCloseRequested += NewWindow_WindowCloseRequested;
}
```

**Comment by jschroedl** [comment]
The NewWindowRequested notification is simply to let your code know that the web page wants a new window. If you take over the processing, it's up to you to create that new window containing a WebView2 and have that new WV2 pick up the navigation using the deferral.  So, it's up to you to create a window and hook up its close event.

**Comment by Dave-Jalpan** [comment]
@joshholmes 

Usually I handled all new window request to open new tab (WebView2). But one of the website need header referrer. 
So for based on suggestions I have written following code but still no luck 
Any suggestion or correction ??? 

```
public class MyWebView2 : WebView2
    {
        internal Microsoft.Web.WebView2.Core.CoreWebView2Deferral Deferral;
        internal Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs EventArgs;
    }

```
```
public MyWebView2 Browser;

private void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
        {

            if(e.Uri.ToUpper().Contains(vURL.ToUpper()))  // Check URL match 
            {
                Browser = GenerateNewWindowWebBrowser();  // Create New MyWebView2 (to maintain header values)
                Browser.Deferral = e.GetDeferral();
                Browser.EventArgs = e;
            }
            else         
            {
               baseTemplate.AddTab(e.Uri, true); // Existing working new tab webView2 creation (using url string) (without extended MyWebView2)
                e.Handled = true;
            }

```

```
public MyWebView2 GenerateNewWindowWebBrowser()
        {
            try
            {
                WebView2 = new MyWebView2();
                browserWorker.ProgressChanged += new ProgressChangedEventHandler(BrowserWorker_ProgressChange);
                browserWorker.DoWork += new DoWorkEventHandler(BrowserWorker_DoWork);
                browserWorker.WorkerReportsProgress = true;
                WebView2.Source = new System.Uri("https://www.google.com");
                WebView2.CoreWebView2InitializationCompleted += WebView2_NewWindowInitialized;  
                WebView2.NavigationCompleted += webBrowser_LoadCompleted;
            }
            catch (Exception ex)
            {
                LoggedExceptionHandling.LogExceptionHandling("Unable to generate web browser", "CustomTabItem.GenerateNewWindowWebBrowser", ex);
            }
            return WebView2;
        }
```

It not reaching WebView2_NewWindowInitialized to complete deferral 

```
private void WebView2_NewWindowInitialized(object sender, Microsoft.Web.WebView2.Core.CoreWebView2InitializationCompletedEventArgs e)
        {
            WebView2.CoreWebView2.NewWindowRequested += CoreWebView2_NewWindowRequested;

            if (e.IsSuccess)
            {
                // Was deferral set in previous instance?
                if (WebView2.Deferral != null)
                {
                    // Set this instance, complete NewWindowRequested and clear deferral

                    CustomTabItem newTab = (CustomTabItem)baseTemplate.GenerateNewItem();
                    newTab.Content = Browser;
                    baseTemplate.currentTabControl.AddTab(newTab, true);

                    WebView2.EventArgs.NewWindow = Browser.CoreWebView2; //Source.CoreWebView2;
                    WebView2.Deferral.Complete();
                    WebView2.Deferral = null;
                    WebView2.EventArgs = null;
                }
            }

        }

```
